Dark theme

cecton opened the pull request:
**_I just learned that webogram is going to be deprecated in favor of https://github.com/evgeny-nadymov/telegram-react Since this version is already accessible and has a dark mode, I closed this PR._**

Following up on #1100

zhukov commented:
Very nice PR. Would be happy to merge if more details are covered in dark.less. Still lots of light colors in UI currently.

zhukov commented:
Also consider using applying style changes with some <html> or <body> class, not attached css. This way we don't need to refresh page on setting change.

nazimjamil commented on app/js/controllers.js:
```diff
@@ -4415,7 +4416,7 @@ angular.module('myApp.controllers', ['myApp.i18n'])
     }
 
-    Storage.get('notify_nodesktop', 'send_ctrlenter', 'notify_volume', 'notify_novibrate', 'notify_nopreview', 'notify_nopush').then(function (settings) {
+    Storage.get('notify_nodesktop', 'send_ctrlenter', 'notify_volume', 'notify_novibrate', 'notify_nopreview', 'notify_nopush', 'appearance_nightmode').then(function (settings) {
```
We should be referring to the theme as 'dark' not 'night'

cecton replied:
done

nazimjamil commented on app/js/controllers.js:
```diff
@@ -4515,6 +4519,17 @@ angular.module('myApp.controllers', ['myApp.i18n'])
+      $scope.toggleNightmode = function () {
+        $scope.appearance.nightmode = !$scope.appearance.nightmode;
+        if (!$scope.appearance.nightmode) {
+          Storage.remove('appearance_nightmode');
```
Perhaps we can store this key in a config somewhere and reference it everywhere else.

cecton replied:
I don't know how to do that and I don't want to get too much into the code base. I propose, if you don't mind and if the code is good enough, to make this change later on master.

nazimjamil commented on app/less/themes/dark.less:
```diff
@@ -0,0 +1,112 @@
+/* Dark theme common css */
```
Maybe a general improvement to the codebase (since we're working on themes) is to add a colours file, that way we can be consistent in how we'er referencing them (e.g. codebase includes both #fff and #ffffff) - this also paves the way for more themes :)

nazimjamil replied:
Looks like we could extend over here.. https://github.com/zhukov/webogram/blob/master/app/less/lib/mixins.less

cecton replied:
I won't go too much into details in this PR as I don't know the code base at all

nazimjamil commented on app/js/controllers.js:
```diff
@@ -4278,6 +4278,7 @@ angular.module('myApp.controllers', ['myApp.i18n'])
 
     $scope.notify = {volume: 0.5}
     $scope.send = {}
+    $scope.appearance = {};
```
Could we call this `theme`?

cecton replied:
done

cecton commented:
I completely forgot about this PR. I did it because I'm one of those users who would love to have a dark theme on webogram and I have been asked in the issue to do it but I have not wrote the code. I think it was as bad idea to ask me to make the PR 😂 I will try to do the changes that have been asked but I know nothing about the code base of webogram.

cecton commented:
@nazimjamil @zhukov @deandreamatias I have fixed the conflict again, renamed some variables as suggested and I believe it is now ready to be merged.

As I said in [this comment](https://github.com/zhukov/webogram/pull/1840#issuecomment-536613419) I'm not the author of this PR. I just made the PR for the greater good of all dark mode users of webogram. I don't really want to get too involved.

Hopefully this PR will meet the minimum quality requested so all users will benefit of the dark mode. I personally love webogram, I'm just not too fan of javascript, hope you understand.

nazimjamil commented:
Hi @cecton! Thank you for making the changes! I actually was keen on this feature and noticed the PR so thought I could help.

IMO it’s good to be merged but am also very happy to make any updates the community suggest.

Looks like the repo is sort of stale right now so I’ll see if I can get in touch with the owners and get this PR merged in!

nazimjamil commented:
Hi @cecton I'm trying to test the feature locally, but when I toggle the setting for night mode, the app refreshes and nothing new seems to be applied. I don't see any errors in the console either - is there a flag elsewhere that I need to switch before I use the feature?

I've been resolving some of my comments that I added to your PR and was going to raise a PR to your branch with the changes :)

cecton commented:
@nazimjamil it's not my code. I have no idea. I expect that in the meanwhile the code evolved so much that the original changes are now not compatible anymore. I will have a look today to help.

cecton commented:
@nazimjamil apparently I f up the renaming of "appearance" to "theme" and "night" to "dark". I really shouldn't have make a PR for someone else in the first place.

I've slightly changed your Dockerfile in the process for my testing purpose but feel free to drop it if you don't like it. It moves the "npm install" to its own layer so you can run `docker build` multiple times and Docker will skip the "npm install" step if nothing changed in the `package{,-lock}.json`

I have tested and it is now working.

cecton commented:
@nazimjamil @zhukov @Feijk Hello o/ Can we finish merging this PR before it gets outdated again?

nazimjamil commented:
Hi @cecton looking at the commit history for master it looks like there has been zero activity since May! Seems to be abandoned by OP and can't identify anyone else with write access 😱

cecton commented:
I just saw a new deployment :grinning: webogram is not dead!

nazimjamil commented:
@zhukov do you need a hand maintaining the repo? Happy to help!

oxr463 commented on app/js/controllers.js:
```diff
@@ -4415,7 +4416,7 @@ angular.module('myApp.controllers', ['myApp.i18n'])
     }
 
-    Storage.get('notify_nodesktop', 'send_ctrlenter', 'notify_volume', 'notify_novibrate', 'notify_nopreview', 'notify_nopush').then(function (settings) {
+    Storage.get('notify_nodesktop', 'send_ctrlenter', 'notify_volume', 'notify_novibrate', 'notify_nopreview', 'notify_nopush', 'theme_darkmode').then(function (settings) {
```
Perhaps this should be named just `theme`, or `active_theme`; that way it is more generic, (for further theming later on).

oxr463 commented on app/js/controllers.js:
```diff
@@ -4433,6 +4434,9 @@ angular.module('myApp.controllers', ['myApp.i18n'])
       $scope.notify.preview = !settings[4]
 
+      // Set default to false
+      $scope.theme.darkmode = settings[6];
```
Again, make this more generic:

```
$scope.theme = settings[6]
```

This should not be a boolean, but rather a string value, e.g., `settings[6] = 'dark'`.

oxr463 commented on app/js/controllers.js:
```diff
@@ -4515,6 +4519,17 @@ angular.module('myApp.controllers', ['myApp.i18n'])
         $rootScope.$broadcast('settings_changed')
       }
+
+      $scope.toggleDarkmode = function () {
```
```
$scope.setTheme = function () {
  if (!$scope.theme) {
    Storage.remove('active_theme');
  } else {
    Storage.set({active_theme: 'default'});
  }
  $rootScope.$broadcast('settings_changed');
  LayoutSwitchService.switchLayout(false);
}
```
This may not work exactly as-is, but you get the idea.

oxr463 commented on app/js/init.js:
```diff
@@ -18,9 +18,10 @@
   $(document.body).addClass(classes.join(' '))
 
-  ConfigStorage.get('layout_selected', 'i18n_locale', function (params) {
+  ConfigStorage.get('layout_selected', 'i18n_locale', 'theme_darkmode', function (params) {
```
`'active_theme'`

oxr463 commented on app/js/init.js:
```diff
@@ -18,9 +18,10 @@
+  ConfigStorage.get('layout_selected', 'i18n_locale', 'theme_darkmode', function (params) {
     var layout = params[0]
     var locale = params[1]
+    var darkmode = params[2]
```
`var theme = params[2]`

oxr463 commented on app/js/init.js:
```diff
@@ -66,6 +67,13 @@
       '<link rel="stylesheet" href="css/' + (Config.Mobile ? 'mobile.css' : 'desktop.css') + '" />'
     )
 
+    // Nightmode CSS injection if found in stored configs
```
```
// Theme CSS injection if found in stored configs
if (active_theme != 'default') {
  $('head').append(
    '<link rel="stylesheet" href="css/' + active_theme + '.css' />'
  );
}
```

oxr463 commented on app/js/locales/de-de.json:
```diff
@@ -66,6 +66,7 @@
     "settings_modal_pushes": "Hintergrund-Benachrichtigungen",
     "settings_modal_message_preview": "Nachrichtenvorschau",
     "settings_modal_sound": "Ton",
+    "settings_modal_dark_mode": "Nacht-Modus",
```
`"settings_modal_active_theme": "Aktiv-Thema",`

Might need to double check since I used Google Translate.

oxr463 commented on app/js/locales/en-us.json:
```diff
@@ -70,6 +70,7 @@
 	"settings_modal_pushes": "Background notifications",
 	"settings_modal_message_preview": "Message preview",
 	"settings_modal_sound": "Sound",
+	"settings_modal_dark_mode": "Night mode",
```
`"settings_modal_active_theme": "Active theme"`

oxr463 commented on app/js/locales/es-es.json:
```diff
@@ -66,6 +66,7 @@
     "settings_modal_message_preview": "Vista previa del mensaje",
     "settings_modal_sound": "Sonido",
+    "settings_modal_dark_mode": "Modo nocturno",
     "settings_modal_enter_send_description_md": "**Intro** - enviar mensaje, **Shift + Intro** - nueva línea",
```
Too lazy to fix this one.

oxr463 commented on app/js/locales/it-it.json:
```diff
@@ -66,6 +66,7 @@
     "settings_modal_pushes": "Notifiche in background",
     "settings_modal_message_preview": "Anteprima messaggio",
     "settings_modal_sound": "Suono",
+    "settings_modal_dark_mode": "Modalità notturna",
```
Too lazy to fix this one.

oxr463 commented on app/js/locales/nl-nl.json:
```diff
@@ -62,6 +62,7 @@
     "settings_modal_vibrate": "Trillen",
     "settings_modal_sounds": "Geluiden",
     "settings_modal_language": "Taal",
+    "settings_modal_dark_mode": "Nachtstand",
```
Too lazy to fix this one.

oxr463 commented on app/js/locales/pt-br.json:
```diff
@@ -66,6 +66,7 @@
     "settings_modal_pushes": "Notificações em segundo plano",
     "settings_modal_message_preview": "Pré-Visualização",
     "settings_modal_sound": "Som",
+    "settings_modal_dark_mode": "Modo noturno",
```
Too lazy to fix this one.

oxr463 commented on app/js/locales/ru-ru.json:
```diff
@@ -62,6 +62,7 @@
     "settings_modal_vibrate": "Вибросигнал",
     "settings_modal_sounds": "Звук",
     "settings_modal_language": "Язык",
+    "settings_modal_dark_mode": "Ночной режим",
```
Too lazy to fix this one.

oxr463 commented on app/partials/desktop/settings_modal.html:
```diff
@@ -73,6 +73,11 @@
           <span class="tg_checkbox_label" my-i18n="settings_modal_message_preview"></span>
         </a>
 
+        <a class="md_modal_section_toggle_wrap tg_checkbox" ng-click="toggleDarkmode()" ng-class="theme.darkmode ? 'tg_checkbox_on' : ''">
```
This would need to be changed to `ng-click="setTheme()" ng-class="theme.active_theme"`, and then `my-i18n="settings_modal_active_theme"`.

Instead of a boolean, we will need to make it accept user input, e.g., `dark`.

cecton commented:
If someone have time to fix all the issues feel free to do it. Otherwise it will wait for a weekend...

oxr463 commented:
> 
> 
> If someone have time to fix all the issues feel free to do it. Otherwise it will wait for a weekend...

We can always just merge it as-is, and then later implement these changes for full theme support.

cecton commented:
> > If someone have time to fix all the issues feel free to do it. Otherwise it will wait for a weekend...
> 
> We can always just merge it as-is, and then later implement these changes for full theme support.

:pray: I'm waiting for this since April, I'm not the author of this code and I'm not too fan of JS. I would be so happy if you do so :sweat_smile:

Coooi commented:
@oxr463  @cecton Please merge this one!! 
So excited about it!

cecton commented:
**_I just learned that webogram is going to be deprecated in favor of https://github.com/evgeny-nadymov/telegram-react Since this version is already accessible and has a dark mode, I closed this PR._**
